content/blog/waddlesplash/2023-04-24_condition_variables.md

```diff
@@ -0,0 +1,416 @@
+
+<!--more-->
+
+## Background
+
+Before delving in to Haiku’s condition variables and how they differ from those of other operating systems, it is first important to understand what they are for, and how they fit in with the other thread-related APIs. (For those already very familiar with atomics and thread synchronization primitives, you can likely skip this section entirely, and jump directly to the section titled “Condition variables”.)
+
+Teams (or in more common terminology, “processes”) are built around *threads*, or more properly, *threads of execution*. Any team will have at least one thread (the “main thread”), and any number of other threads. On systems with multiple CPU cores (and/or “hyperthreads”), multiple threads will be running at once; and even on single-core systems, the order in which threads run will not always be the same, since it may depend on disk I/O, system load, etc. Thus, thread operations need to be *synchronized.*
+
+### Atomics
+
+Consider if I have a variable which I want to add 1 to.
+```cpp
+int32 gSharedInteger = /* initializer... */;
+void add() {
+	int32 localInteger = gSharedInteger;
+	localInteger = localInteger + 1;
+	gSharedInteger = localInteger;
+}
+```
+What happens if `add` is invoked for the first time simultaneously in two separate threads? Will `sharedInteger` become `2`, or (if the invocations really occur at exactly the same instant), will it become `1`, since the threads will have “raced” against each other?
+
+To solve this problem, ISAs (instruction set architectures) include _atomic operations_, so called because they act like one operation, not distinct ones. In our `add` function, we performed 3 distinct operations: we fetched the value, added `1` to it, and then set the value. With an atomic operation, these 3 operations become 1 single operation (for all intents and purposes of the programmer, anyway), and so what order threads run them in is irrelevant.
+
+Different compilers, operating systems, etc. expose such ISA-level atomic operations in different ways. On Haiku, we expose them through a set of functions beginning with `atomic_` which operate on `int32*`s (and there is also an equivalent set of functions for `int64*`s.) If we use one of them to make our earlier example thread-safe, it will look like this:
+```cpp
+void add() {
+	atomic_add(gSharedInteger, 1);
+}
+```
+(As a bonus, `atomic_add` also returns the value `gSharedInteger` had before `1` was added to it: so it is really a “fetch and add” operation, not just an “add” operation.)
+
+There are a variety of different atomic operations, and many ways to use them. But it is important to keep in mind that these are a ‘low-level’ primitive, and when using them, there are a lot of considerations that must be taken into account: when using multiple atomic operations in a row, interesting things can happen with memory ordering, for example.
```

Comment:
a link to the documentation for atomic operations in the haiku book (or if there is none, to the header file where we define them) would be nice here